File: July 2025/07 July/Solution.cpp
// Title: Maximum Number of Events That Can Be Attended




#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maxEvents(vector<vector<int>>& events) {
        sort(events.begin(), events.end());
        priority_queue<int, vector<int>, greater<int>> minHeap;
        int day = 0, idx = 0, n = events.size(), result = 0;

        while (!minHeap.empty() || idx < n) {
            if (minHeap.empty()) day = events[idx][0];

            while (idx < n && events[idx][0] <= day) {
                minHeap.push(events[idx][1]);
                idx++;
            }

            minHeap.pop();
            result++;
            day++;

            while (!minHeap.empty() && minHeap.top() < day) minHeap.pop();
        }

        return result;
    }
};

File: July 2025/07 July/Solution.py
// Title: Maximum Number of Events That Can Be Attended




import heapq

class Solution:
    def maxEvents(self, events: list[list[int]]) -> int:
        events.sort()
        min_heap = []
        day = 0
        idx = 0
        result = 0
        n = len(events)

        while idx < n or min_heap:
            if not min_heap:
                day = events[idx][0]

            while idx < n and events[idx][0] <= day:
                heapq.heappush(min_heap, events[idx][1])
                idx += 1

            heapq.heappop(min_heap)
            result += 1
            day += 1

            while min_heap and min_heap[0] < day:
                heapq.heappop(min_heap)

        return result